### PR TITLE
Add WHATWG transform streams

### PR DESCRIPTION
**Pull request by timostamm**
This adds transform streams for compression, serialization, and enveloping to @bufbuild/connect-core. We will use this feature to streamline the handler and client implementations in @bufbuild/connect-node.

### CODE REVIEW

File: packages/connect-core/src/transform-stream-story.spec.ts [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+import { ConnectError } from "./connect-error.js";
+import { Code } from "./code.js";
+
+node16WhatwgStreamPolyfill();
+
+// These tests aim to model the usage of transform streams in clients and servers.
+// Note that the tests were written as a proof of concept, and the coverage is
+// incomplete (cancellation, backpressure, error handling, etc.).
+describe("full story", function () {
+  const readMaxBytes = 0xffffffff; // zlib caps maxOutputLength at this value
```
Comment (smaye81) [inline packages/connect-core/src/transform-stream-story.spec.ts:42]: Is it worth reading the max bytes value from here?:  packages/connect-node/src/private/validate-read-max-bytes-option.ts

File: packages/connect-core/src/transform-stream.ts [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    transform(chunk, controller) {
+      let data: Uint8Array;
+      let flags = 0;
+      if (chunk.end) {
+        flags = flags | endStreamFlag;
+        try {
+          data = endSerialization.serialize(chunk.value);
+        } catch (e) {
+          return controller.error(
+            new ConnectError("failed to serialize end", Code.Internal)
```
Comment (smaye81) [inline packages/connect-core/src/transform-stream.ts:55]: Should we specify `e` as the `cause` here (and a bunch of places below)?

**Review by smaye81 — APPROVED** [review]

File: packages/connect-core/src/transform-stream-story.spec.ts (hunk 1 above)
Comment (timostamm) [reply packages/connect-core/src/transform-stream-story.spec.ts:42]: We can't, this is in connect-core.

File: packages/connect-core/src/transform-stream.ts (hunk 2 above)
Comment (timostamm) [reply packages/connect-core/src/transform-stream.ts]: Yes, good shout. Updated.

**Review by gilwong00 — APPROVED** [review]

**Pull request merged**